### PR TITLE
Support extra stuff in the Content-Type header via Mime registration

### LINKED ISSUE #36993: Fragment content type issues in Rails 6.0
We're seeing an issue in Rails 6.0 that we actually found previously in 5.2 but didn't follow up with on 6.0. In #35549 the content type parser changed and now we're seeing the wrong headers getting returned which results in our error messages displaying as "Something went wrong" rather than "Username is already taken". The error is visible in the response on the browser console but doesn't make it to the front-end.

Here are the headers in 5.2:

63308096-e17d7a00-c2a5-11e9-93b1-bb5bb6d5b2c1

And 6.0

63308102-e80bf180-c2a5-11e9-8ce1-7ac155849334-3

And here's a repro script @tenderlove wrote up demonstrating the issue:

```ruby
# frozen_string_literal: true

require "abstract_unit"
require "active_support/log_subscriber/test_helper"

class RespondToController < ActionController::Base
  layout :set_layout

  def my_html_fragment
    respond_to do |type|
      type.html_fragment { render body: "neat" }
    end
  end
end

class RespondToControllerTest < ActionController::TestCase
  def setup
    super
    @request.host = "www.example.com"
    Mime::Type.register("text/html; fragment", :html_fragment)
  end

  def teardown
    super
    Mime::Type.unregister(:html_fragment)
  end

  def test_html_fragment
    @request.accept = "text/html; fragment"
    get :my_html_fragment
    assert_equal "text/html; fragment; charset=utf-8", @response.headers["Content-Type"]
    assert_equal "neat", @response.body
  end
end
```

We're not sure exactly how to fix this yet but wanted to open an issue until we figure it out.

cc/ @tenderlove @jhawthorn @cpruitt @alimi @rafaelfranca @kamipo 
cc original author/ @r7kamura

### PR DESCRIPTION
**Pull request by tenderlove**
I'm not sure a better way to describe this, but we need to set extra
stuff in the Content-Type header and we're doing it via Mime
registration so that we can have the "html_fragment" method on the
respond_to type.  This worked in Rails 5.2, but it's broken in 6.0

In the future, I think we should probably give `register` a block that
takes the response object so we can do more interesting and complex
stuff without doing weird string manipulation.  But this patch fixes the
issue we have today.

Fixes #36993

### CODE REVIEW

**Comment by tenderlove** [comment]
cc @r7kamura

**Review by eileencodes — APPROVED** [review]

**Comment by eileencodes** [comment]
Backported to 6-0-stable in https://github.com/rails/rails/commit/d5205571d5dfbb1044e1d9a0923859d0d8070616

**Comment by rafaelfranca** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: actionpack/lib/action_dispatch/http/response.rb [hunk 1]
```diff
@@ -419,14 +419,14 @@ def cookies
     end
 
   private
-    ContentTypeHeader = Struct.new :mime_type, :charset
-    NullContentTypeHeader = ContentTypeHeader.new nil, nil
+    ContentTypeHeader = Struct.new :mime_type, :extra, :charset
+    NullContentTypeHeader = ContentTypeHeader.new nil, nil, nil
```
Comment (abhaynikam) [inline actionpack/lib/action_dispatch/http/response.rb]: `NullContentTypeHeader` is not used anywhere. Can that be removed? 

I tried removing it and running the spec everything worked fine for me. Just not sure if I should raise a PR or not.

File: actionpack/lib/action_dispatch/http/response.rb [hunk 2]
```diff
@@ -419,14 +419,14 @@ def cookies
...
 
     def parse_content_type(content_type)
       if content_type && match = CONTENT_TYPE_PARSER.match(content_type)
-        ContentTypeHeader.new(match[:type], match[:charset])
+        ContentTypeHeader.new(match[:type], match[:extra], match[:charset])
       else
-        NullContentTypeHeader
+        ContentTypeHeader.new(content_type, nil)
```
Comment (kamipo) [inline actionpack/lib/action_dispatch/http/response.rb]: Is this change required for GitHub?
Looks like this change is not tested.

Comment (tenderlove) [reply actionpack/lib/action_dispatch/http/response.rb]: I don't think so.  I think we can remove the change (you're right, it seems not tested)
